File: kaolin/csrc/render/dibr.cpp
// Copyright (c) 2019-2020, NVIDIA CORPORATION. All rights reserved.
  
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <ATen/ATen.h>

#include "../check.h"

#define CHECK_INPUT(x) CHECK_CUDA(x); CHECK_CONTIGUOUS(x)
#define CHECK_DIM4(x, b, h, w, d) TORCH_CHECK((x.dim() == 4) && (x.size(0) == b) && (x.size(1) == h) && (x.size(2) == w) && (x.size(3) == d), #x " must be same im size")
#define CHECK_DIM3(x, b, f, d) TORCH_CHECK((x.dim() == 3) && (x.size(0) == b) && (x.size(1) == f) && (x.size(2) == d), #x " must be same point size")
#define CHECK_DIM2(x, b, d) TORCH_CHECK((x.dim() == 2) && (x.size(0) == b) && (x.size(1) == d), #x " wrong size")
#define CHECK_DIM1(x, b) TORCH_CHECK((x.dim() == 1) && (x.size(0) == b), #x " wrong size")

namespace kaolin {

#ifdef WITH_CUDA

void packed_rasterize_forward_cuda_kernel_launcher(
    at::Tensor face_vertices_z,
    at::Tensor face_vertices_image,
    at::Tensor face_bboxes,
    at::Tensor face_features,
    at::Tensor first_idx_face_per_mesh,
    at::Tensor selected_face_idx,
    at::Tensor output_weights,
    at::Tensor interpolated_features,
    float multiplier);

void rasterize_backward_cuda_kernel_launcher(
    at::Tensor grad_interpolated_features,
    at::Tensor grad_improb_bxhxwx1,
    at::Tensor interpolated_features,
    at::Tensor improb_bxhxwx1,
    at::Tensor selected_face_idx,
    at::Tensor output_weights,
    at::Tensor probface_bxhxwxk,
    at::Tensor probcase_bxhxwxk,
    at::Tensor probdis_bxhxwxk,
    at::Tensor face_vertices_image,
    at::Tensor face_features,
    at::Tensor grad_face_vertices_image,
    at::Tensor grad_face_features,
    at::Tensor grad_points2dprob_bxfx6,
    int multiplier,
    int sigmainv);

void generate_soft_mask_cuda_kernel_launcher(
    at::Tensor face_vertices_image,
    at::Tensor face_bboxes,
    at::Tensor selected_face_idx,
    at::Tensor probface_bxhxwxk,
    at::Tensor probcase_bxhxwxk,
    at::Tensor probdis_bxhxwxk,
    at::Tensor improb_bxhxwx1,
    float multiplier,
    float sigmainv);

#endif  // WITH_CUDA

void packed_rasterize_forward_cuda(
    at::Tensor face_vertices_z,         // depth of face_vertices in camera ref.
    at::Tensor face_vertices_image,     // x,y coordinates of face_vertices in camera plan.
    at::Tensor face_bboxes,             // bounding boxes of faces in camera plan.
    at::Tensor face_features,           // features to interpolate.
    at::Tensor first_idx_face_per_mesh, // indices of the first face of each mesh (+ last_idx + 1).
    at::Tensor selected_face_idx,       // output: indices within each mesh of selected faces.
    at::Tensor output_weights,          // output: the weights used for interpolation.
    at::Tensor interpolated_features,   // output: the interpolated face features.
    float multiplier) {                 // coordinates multiplier used to improve numerical precision.

  CHECK_INPUT(face_vertices_z);
  CHECK_INPUT(face_vertices_image);
  CHECK_INPUT(face_bboxes);
  CHECK_INPUT(face_features);
  // Usually for packed representation first_idx is on CPU
  // but here it's on GPU because or the way we computed it on DIBRasterization
  CHECK_INPUT(first_idx_face_per_mesh);

  CHECK_INPUT(selected_face_idx);
  CHECK_INPUT(output_weights);

  CHECK_INPUT(interpolated_features);

  int num_faces = face_vertices_z.size(0);
  int batch_size = interpolated_features.size(0);
  int height = interpolated_features.size(1);
  int width = interpolated_features.size(2);
  int num_features = interpolated_features.size(3);

  CHECK_DIM2(face_vertices_z, num_faces, 3);
  CHECK_DIM3(face_vertices_image, num_faces, 3, 2);
  CHECK_DIM2(face_bboxes, num_faces, 4);
  CHECK_DIM3(face_features, num_faces, 3, num_features);
  CHECK_DIM1(first_idx_face_per_mesh, batch_size + 1)

  CHECK_DIM3(selected_face_idx, batch_size, height, width);
  CHECK_DIM4(output_weights, batch_size, height, width, 3);
  CHECK_DIM4(interpolated_features, batch_size, height, width, num_features);

#if WITH_CUDA
  packed_rasterize_forward_cuda_kernel_launcher(
      face_vertices_z, face_vertices_image, face_bboxes,
      face_features, first_idx_face_per_mesh, selected_face_idx,
      output_weights, interpolated_features, multiplier);
  return;
#else
  AT_ERROR("packed_rasterize_forward not built with CUDA");
#endif  // WITH_CUDA
}

void generate_soft_mask_cuda(
    at::Tensor face_vertices_image, // x,y coordinates of face_vertices in camera plan.
    at::Tensor face_bboxes,         // enlarged bounding boxes of faces in camera plan.
    at::Tensor selected_face_idx,   // indices within each mesh of selected faces.
    at::Tensor probface_bxhxwxk,    // k nearby faces that influences it, 0 for void, per pixel
    at::Tensor probcase_bxhxwxk,    // type of distance from the face to the pixel
                                    // (0, 1, 2 means minimal distance from pixel to edges
				    // 3, 4, 5 means minimal distance from pixel to vertices), per pixel
    at::Tensor probdis_bxhxwxk,     // distance probability, per pixel
    at::Tensor improb_bxhxwx1,      // soft mask, per pixel
    float multiplier,               // coordinates multiplier used to improve numerical precision.
    float sigmainv) {               // smoothness term for soft mask, the higher the shaper, range is (1/3e-4, 1/3e-5). default: 7000

  CHECK_INPUT(face_vertices_image);
  CHECK_INPUT(face_bboxes);
  CHECK_INPUT(selected_face_idx);

  CHECK_INPUT(probface_bxhxwxk);
  CHECK_INPUT(probcase_bxhxwxk);
  CHECK_INPUT(probdis_bxhxwxk);

  CHECK_INPUT(improb_bxhxwx1);

  int batch_size = face_vertices_image.size(0);
  int num_faces = face_vertices_image.size(1);
  int height = improb_bxhxwx1.size(1);
  int width = improb_bxhxwx1.size(2);

  int knum = probface_bxhxwxk.size(3);

  CHECK_DIM4(face_vertices_image, batch_size, num_faces, 3, 2);
  CHECK_DIM3(face_bboxes, batch_size, num_faces, 4);

  CHECK_DIM3(selected_face_idx, batch_size, height, width);

  CHECK_DIM4(probface_bxhxwxk, batch_size, height, width, knum);
  CHECK_DIM4(probcase_bxhxwxk, batch_size, height, width, knum);
  CHECK_DIM4(probdis_bxhxwxk, batch_size, height, width, knum);

  CHECK_DIM3(improb_bxhxwx1, batch_size, height, width);

#if WITH_CUDA
  generate_soft_mask_cuda_kernel_launcher(
      face_vertices_image, face_bboxes, selected_face_idx,
      probface_bxhxwxk, probcase_bxhxwxk, probdis_bxhxwxk,
      improb_bxhxwx1, multiplier, sigmainv);
  return;
#else
  AT_ERROR("generate_soft_mask not built with CUDA");
#endif
}

void rasterize_backward_cuda(
    at::Tensor grad_interpolated_features, // gradients from interpolated_features.
    at::Tensor grad_improb_bxhxwx1,        // gradients from improb_bxhxwx1.
    at::Tensor interpolated_features,      // the interpolated face features.
    at::Tensor improb_bxhxwx1,             // soft mask, per pixel.
    at::Tensor selected_face_idx,          // indices within each mesh of selected faces.
    at::Tensor output_weights,             // the weights used for interpolation.
    at::Tensor probface_bxhxwxk,           // k nearby faces that influences it, 0 for void, per pixel.
    at::Tensor probcase_bxhxwxk,           // type of distance from the face to the pixel
                                           // (0, 1, 2 means minimal distance from pixel to edges
                                           // 3, 4, 5 means minimal distance from pixel to vertices), per pixel.
    at::Tensor probdis_bxhxwxk,            // distance probability, per pixel.
    at::Tensor face_vertices_image,        // x,y coordinates of face_vertices in camera plan.
    at::Tensor face_features,              // features to interpolate.
    at::Tensor grad_face_vertices_image,   // output: gradients to face_vertices_image from the interpolated features
    at::Tensor grad_face_features,         // outputs: gradients to face_features
    at::Tensor grad_points2dprob_bxfx6,    // outputs: gradients to face_vertices_image from the soft mask
    int multiplier,                        // coordinates multiplier used to improve numerical precision.
    int sigmainv) {                        // smoothness term for soft mask, the higher the shaper, range is (1/3e-4, 1/3e-5). default: 7000

  CHECK_INPUT(grad_interpolated_features);
  CHECK_INPUT(grad_improb_bxhxwx1);
  CHECK_INPUT(interpolated_features);
  CHECK_INPUT(improb_bxhxwx1);
  CHECK_INPUT(selected_face_idx);
  CHECK_INPUT(output_weights);

  CHECK_INPUT(probface_bxhxwxk);
  CHECK_INPUT(probcase_bxhxwxk);
  CHECK_INPUT(probdis_bxhxwxk);

  CHECK_INPUT(face_vertices_image);
  CHECK_INPUT(face_features);
  CHECK_INPUT(grad_face_vertices_image);
  CHECK_INPUT(grad_face_features);
  CHECK_INPUT(grad_points2dprob_bxfx6);

  int bnum = grad_interpolated_features.size(0);
  int height = grad_interpolated_features.size(1);
  int width = grad_interpolated_features.size(2);
  int dnum = grad_interpolated_features.size(3);
  int fnum = grad_face_vertices_image.size(1);
  int knum = probface_bxhxwxk.size(3);

  CHECK_DIM4(grad_interpolated_features, bnum, height, width, dnum);
  CHECK_DIM3(grad_improb_bxhxwx1, bnum, height, width);

  CHECK_DIM4(interpolated_features, bnum, height, width, dnum);
  CHECK_DIM3(improb_bxhxwx1, bnum, height, width);

  CHECK_DIM3(selected_face_idx, bnum, height, width);
  CHECK_DIM4(output_weights, bnum, height, width, 3);

  CHECK_DIM4(probface_bxhxwxk, bnum, height, width, knum);
  CHECK_DIM4(probface_bxhxwxk, bnum, height, width, knum);
  CHECK_DIM4(probdis_bxhxwxk, bnum, height, width, knum);

  CHECK_DIM4(face_vertices_image, bnum, fnum, 3, 2);
  CHECK_DIM4(face_features, bnum, fnum, 3, dnum);
  CHECK_DIM4(grad_face_vertices_image, bnum, fnum, 3, 2);
  CHECK_DIM4(grad_face_features, bnum, fnum, 3, dnum);
  CHECK_DIM4(grad_points2dprob_bxfx6, bnum, fnum, 3, 2);

#if WITH_CUDA
  rasterize_backward_cuda_kernel_launcher(grad_interpolated_features, grad_improb_bxhxwx1,
      interpolated_features, improb_bxhxwx1, selected_face_idx, output_weights,
      probface_bxhxwxk, probcase_bxhxwxk, probdis_bxhxwxk, face_vertices_image,
      face_features, grad_face_vertices_image, grad_face_features,
      grad_points2dprob_bxfx6, multiplier, sigmainv);
  return;
#else
  AT_ERROR("rasterize_backward not built with CUDA");
#endif  // WITH_CUDA
}

}  // namespace kaolin

#undef CHECK_DIM1
#undef CHECK_DIM2
#undef CHECK_DIM3
#undef CHECK_DIM4

File: kaolin/render/mesh/dibr.py
# Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES.
# All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import torch
from torch.autograd import Function

from kaolin import _C
from .rasterization import rasterize, _legacy_to_opengl, nvdiff, _get_nvdiff_glctx

__all__ = [
    "dibr_soft_mask",
    "dibr_rasterization",
]

class DibrSoftMaskCuda(Function):
    @staticmethod
    def forward(ctx, face_vertices_image, selected_face_idx,
                sigmainv, boxlen, knum, multiplier):
        face_vertices_image = face_vertices_image.contiguous()
        face_vertices_image = face_vertices_image * multiplier
        selected_face_idx = selected_face_idx.contiguous()
        points_min = torch.min(face_vertices_image, dim=-2)[0]
        points_max = torch.max(face_vertices_image, dim=-2)[0]
        face_large_bboxes = torch.cat([
            points_min - boxlen * multiplier,
            points_max + boxlen * multiplier
        ], dim=-1)
        soft_mask, close_face_prob, close_face_idx, close_face_dist_type = \
            _C.render.mesh.dibr_soft_mask_forward_cuda(
                face_vertices_image,
                face_large_bboxes.contiguous(),
                selected_face_idx,
                sigmainv,
                knum,
                multiplier
            )
        ctx.multiplier = multiplier
        ctx.sigmainv = sigmainv
        ctx.save_for_backward(
            soft_mask, face_vertices_image, selected_face_idx,
            close_face_prob, close_face_idx, close_face_dist_type
        )
        return soft_mask

    @staticmethod
    def backward(ctx, grad_soft_mask):
        soft_mask, face_vertices_image, selected_face_idx, close_face_prob, \
            close_face_idx, close_face_dist_type = ctx.saved_tensors
        multiplier = ctx.multiplier
        sigmainv = ctx.sigmainv
        grad_face_vertices_image = _C.render.mesh.dibr_soft_mask_backward_cuda(
            grad_soft_mask.contiguous(),
            soft_mask,
            selected_face_idx,
            close_face_prob,
            close_face_idx,
            close_face_dist_type,
            face_vertices_image,
            sigmainv,
            multiplier)
        return grad_face_vertices_image, None, None, None, None, None

def dibr_soft_mask(face_vertices_image, selected_face_idx,
                   sigmainv=7000, boxlen=0.02, knum=30, multiplier=1000.):
    r"""Compute a soft mask generally used with :func:`kaolin.metrics.render.mask_iou`
    to compute a silhouette loss, as defined by *Chen, Wenzheng, et al.* in
    `Learning to Predict 3D Objects with an Interpolation-based Differentiable Renderer`_ Neurip 2019.

    Args:
        face_vertices_image (torch.Tensor):
            2D positions of the face vertices on image plane,
            of shape :math:`(\text{batch_size}, \text{num_faces}, 3, 2)`,
            Note that ``face_vertices_camera`` is projected on image plane (z=-1)
            and forms ``face_vertices_image``.
            The coordinates of face_vertices_image are between :math:`[-1, 1]`,
            which corresponds to normalized image pixels.
        selected_face_idx (torch.LongTensor):
            Rendered face index,
            of shape :math:`(\text{batch_size}, \text{height}, \text{width})`.
            See 2nd returned value from :func:`kaolin.render.mesh.rasterize`.
        sigmainv (float):
            Smoothness term for computing the softmask, the higher the sharper.
            The recommended range is :math:`[1/3e-4, 1/3e-5]`. Defaut: 7000.
        boxlen (float):
            Margin over bounding box of faces which will threshold which pixels
            will be influenced by the face. The value should be adapted to sigmainv,
            to threshold values close to 0. The recommended range is [0.05, 0.2].
            Default: 0.02.
        knum (int):
            Maximum number of faces that can influence one pixel.
            The value should be adapted to boxlen, to avoid missing faces.
            The recommended range is [20, 100]. Default: 30.
        multiplier (float):
            To avoid numerical issue,
            we internally enlarge the 2d coordinates by a multiplier.
            Default: 1000.
    Returns:
        (torch.FloatTensor):
        The soft mask, of shape :math:`(\text{batch_size}, \text{height}, \text{width})`.

    .. _Learning to Predict 3D Objects with an Interpolation-based Differentiable Renderer:
        https://arxiv.org/abs/1908.01210
    """
    return DibrSoftMaskCuda.apply(face_vertices_image, selected_face_idx,
                                  sigmainv, boxlen, knum, multiplier)

def dibr_rasterization(height, width, face_vertices_z, face_vertices_image,
                       face_features, face_normals_z, sigmainv=7000,
                       boxlen=0.02, knum=30, multiplier=None, eps=None,
                       rast_backend='cuda'):
    r"""Fully differentiable DIB-R renderer implementation,
    that renders 3D triangle meshes with per-vertex per-face features to
    generalized feature "images", soft foreground masks, and face index maps.

    Args:
        height (int): the size of rendered images.
        width (int): the size of rendered images.
        face_vertices_z (torch.FloatTensor):
            3D points depth (z) value of the face vertices in camera coordinate,
            of shape :math:`(\text{batch_size}, \text{num_faces}, 3)`.
        face_vertices_image (torch.FloatTensor):
            2D positions of the face vertices on image plane,
            of shape :math:`(\text{batch_size}, \text{num_faces}, 3, 2)`,
            Note that ``face_vertices_camera`` is projected on image plane (z=-1)
            and forms ``face_vertices_image``.
            The coordinates of face_vertices_image are between :math:`[-1, 1]`,
            which corresponds to normalized image pixels.
        face_features (torch.FloatTensor or list of torch.FloatTensor):
            Features (per-vertex per-face) to be drawn,
            of shape :math:`(\text{batch_size}, \text{num_faces}, 3, \text{feature_dim})`,
            feature is the features dimension,
            for instance with vertex colors num_features=3 (R, G, B),
            and texture coordinates num_features=2 (X, Y),
            or a list of num_features,
            of shapes :math:`(\text{batch_size}, \text{num_faces}, 3, \text{feature_dim[i]})`
        face_normals_z (torch.FloatTensor):
            Normal directions in z axis, of shape :math:`(\text{batch_size}, \text{num_faces})`,
            only faces with normal z >= 0 will be drawn.
        sigmainv (float):
            Smoothness term for computing the softmask, the higher the sharper.
            The recommended range is :math:`[1/3e-4, 1/3e-5]`. Defaut: 7000.
        boxlen (float):
            Margin over bounding box of faces which will threshold which pixels
            will be influenced by the face. The value should be adapted to sigmainv,
            to threshold values close to 0. The recommended range is [0.05, 0.2].
            Default: 0.02.
        knum (int):
            Maximum number of faces that can influence one pixel.
            The value should be adapted to boxlen, to avoid missing faces.
            The recommended range is [20, 100]. Default: 30.
        multiplier (float):
            To avoid numerical issue,
            we internally enlarge the 2d coordinates by a multiplier.
            Default: 1000.
        eps (float):
            Epsilon value used to normalize barycentric weights in rasterization.
            Especially matter with small triangles,
            to increase or decrease in case of exploding or vanishing gradient.
            Ignored if ``backend`` is 'nvdiffrast'.
            Default: 1e-8.
        backend (string):
            Backend used for the rasterization, can be ['cuda', 'nvdiffrast', nvdiffrast_fwd'].
            'nvdiffrast_fwd' is using `nvdiffrast library` for the forward pass only
            and kaolin's custom Op for backward pass.

    Returns:
        (torch.Tensor, torch.Tensor, torch.LongTensor):

        - The rendered features of shape
          :math:`(\text{batch_size}, \text{height}, \text{width}, \text{num_features})`,
          if `face_features` is a list of torch.FloatTensor, return of torch.FloatTensor,
          of shapes :math:`(\text{batch_size}, \text{height}, \text{width}, \text{num_features[i]})`.
        - The rendered soft mask, of shape :math:`(\text{batch_size}, \text{height}, \text{width})`.
          It is generally used with :func:`kaolin.metrics.render.mask_iou` to compute the silhouette loss.
        - The rendered face index, -1 is None,
          of shape :math:`(\text{batch_size}, \text{height}, \text{width})`.
    """
    interpolated_features, face_idx = rasterize(
        height, width,
        face_vertices_z,
        face_vertices_image,
        face_features,
        face_normals_z >= 0.,
        multiplier,
        eps,
        rast_backend
    )
    _multiplier = 1000. if multiplier is None else multiplier
    soft_mask = dibr_soft_mask(
        face_vertices_image,
        face_idx,
        sigmainv,
        boxlen,
        knum,
        _multiplier
    )
    return interpolated_features, soft_mask, face_idx
